language/move-prover/bytecode/src/read_write_set_analysis.rs

```diff
@@ -879,3 +842,67 @@ impl Default for ReadWriteSetState {
         }
     }
 }
+
+// =================================================================================================
+// Converting Infer Result into standalone types
+
+impl Offset {
+    pub fn normalize(&self, env: &GlobalEnv) -> Option<RWOffset> {
+        Some(match self {
+            Offset::Field(idx) => RWOffset::Field(*idx),
+            Offset::VectorIndex => RWOffset::VectorIndex,
+            Offset::Global(s) => RWOffset::Global(s.get_type().into_struct_type(env)?),
+        })
+    }
+}
+impl AccessPath {
+    pub fn normalize(&self, env: &GlobalEnv) -> Option<Vec<RWAccessPath>> {
+        let (roots, access_opt) = match self.root() {
+            Root::Formal(idx) => (vec![RWRoot::Formal(*idx)], None),
+            Root::Global(key) => (
+                key.address()
+                    .get_concrete_addresses()
+                    .into_iter()
+                    .map(RWRoot::Const)
+                    .collect(),
+                key.struct_type().get_type().into_struct_type(env),
+            ),
+            _ => return None,
+        };
+        let mut offsets = access_opt
+            .clone()
+            .map_or_else(|| vec![], |struct_ty| vec![RWOffset::Global(struct_ty)]);
+
+        offsets.append(
+            &mut self
+                .offsets()
+                .iter()
+                .map(|offset| offset.normalize(env))
+                .collect::<Option<Vec<_>>>()?,
+        );
+
+        Some(
+            roots
+                .into_iter()
+                .map(|root| RWAccessPath {
+                    root,
+                    offsets: offsets.clone(),
+                })
+                .collect(),
+        )
+    }
+}
+
+impl ReadWriteSetState {
+    pub fn normalize(&self, env: &GlobalEnv) -> Option<ReadWriteSet> {
```

Comment:
I don't think this ever returns `None`.

Reply:
You are right. This is because the `iter_path` here doesn't contain an early termniation option. Do you suggest I go implement it in the `iter_path` or I can implement a flag to determine if an error occur here.

Reply:
My dream is to eventually replace the hacky `iter_paths` with a proper `impl Iter<Item=(AccessPath,Option<T>)>` (or similar). That way, we can use the normal `find` to do normal termination. But it's tricky to implement the tree traversal in a non-recursive fashion...

I think it's safe to do an `expect` here + return `ReadWriteSet`--normalization should never fail unless there's a bug in the analysis or `GlobalEnv`.

Reply:
Added `expect` here and removed `Option`